### PR TITLE
[ISSUE #10992] Add unit test for shardingsphere-sql-parser-engine 

### PR DESCRIPTION
**Pull request by zzq1314zll**
for #10992
 

- [x] SQLParserEngine
- [ ] SQLVisitorEngine
- [x] SQLParserFactory

### CODE REVIEW

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * SQL parser engine test.
+ *
+ * @author zzq
+ */
+public class SQLParserEngineTest {
```
Comment (soulasuna) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:42]: No annotations are required for test classes.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    
+    @Test
+    public void assertParse() throws NoSuchFieldException, IllegalAccessException {
+        SQLParserExecutor sqlParserExecutor = mock(SQLParserExecutor.class);
+        when(sqlParserExecutor.parse(SQL)).thenReturn(mock(ParseTree.class));
+        
+        LoadingCache<String, ParseTree> parseTreeCache = CacheBuilder.newBuilder().softValues().initialCapacity(128)
+                .maximumSize(1024).concurrencyLevel(4).build(new CacheLoader<String, ParseTree>() {
+                    
+                    @ParametersAreNonnullByDefault
```
Comment (soulasuna) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:54]: Modify the extra blank lines.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                    @Override
+                    public ParseTree load(final String sql) {
+                        return sqlParserExecutor.parse(sql);
+                    }
+                });
+
```
Comment (soulasuna) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:60]: Modify the extra blank lines.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                    
+                    @ParametersAreNonnullByDefault
+                    @Override
+                    public ParseTree load(final String sql) {
+                        return sqlParserExecutor.parse(sql);
+                    }
+                });
+        
+        SQLParserEngine sqlParserEngine = new SQLParserEngine("H2");
+
```
Comment (soulasuna) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:62]: Modify the extra blank lines.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        
+        sqlParserEngine.parse(SQL, true);
+        verify(sqlParserExecutor, times(1)).parse(SQL);
+        sqlParserEngine.parse(SQL, true);
+        verify(sqlParserExecutor, times(1)).parse(SQL);
+        sqlParserEngine.parse(SQL, false);
+        verify(sqlParserExecutor, times(2)).parse(SQL);
+    }
+    
+}
```
Comment (soulasuna) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:88]: Modify the extra blank lines.
Leave a blank line on the last line

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        modifiersField2.setAccessible(true);
+        modifiersField2.setInt(parseTreeCacheField, sqlParserExecutorFiled.getModifiers() & ~Modifier.FINAL);
+        
+        sqlParserExecutorFiled.setAccessible(true);
+        parseTreeCacheField.setAccessible(true);
+        
+        sqlParserExecutorFiled.set(sqlParserEngine, sqlParserExecutor);
+        parseTreeCacheField.set(sqlParserEngine, parseTreeCache);
+        
+        sqlParserEngine.parse(SQL, true);
```
Comment (soulasuna) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:80]: Modify the extra blank lines.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        
+        Field modifiersField = Field.class.getDeclaredField("modifiers");
+        modifiersField.setAccessible(true);
+        modifiersField.setInt(sqlParserExecutorFiled, sqlParserExecutorFiled.getModifiers() & ~Modifier.FINAL);
+        
+        Field modifiersField2 = Field.class.getDeclaredField("modifiers");
+        modifiersField2.setAccessible(true);
+        modifiersField2.setInt(parseTreeCacheField, sqlParserExecutorFiled.getModifiers() & ~Modifier.FINAL);
+        
+        sqlParserExecutorFiled.setAccessible(true);
```
Comment (soulasuna) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:74]: Modify the extra blank lines.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                        return sqlParserExecutor.parse(sql);
+                    }
+                });
+        
+        SQLParserEngine sqlParserEngine = new SQLParserEngine("H2");
+        
+        Field sqlParserExecutorFiled = sqlParserEngine.getClass().getDeclaredField("sqlParserExecutor");
+        Field parseTreeCacheField = sqlParserEngine.getClass().getDeclaredField("parseTreeCache");
+        
+        Field modifiersField = Field.class.getDeclaredField("modifiers");
```
Comment (soulasuna) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:66]: Modify the extra blank lines.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        modifiersField.setAccessible(true);
+        modifiersField.setInt(sqlParserExecutorFiled, sqlParserExecutorFiled.getModifiers() & ~Modifier.FINAL);
+        
+        Field modifiersField2 = Field.class.getDeclaredField("modifiers");
+        modifiersField2.setAccessible(true);
+        modifiersField2.setInt(parseTreeCacheField, sqlParserExecutorFiled.getModifiers() & ~Modifier.FINAL);
+
```
Comment (soulasuna) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:73]: Modify the extra blank lines.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        SQLParserEngine sqlParserEngine = new SQLParserEngine("H2");
+        
+        Field sqlParserExecutorFiled = sqlParserEngine.getClass().getDeclaredField("sqlParserExecutor");
+        Field parseTreeCacheField = sqlParserEngine.getClass().getDeclaredField("parseTreeCache");
+        
+        Field modifiersField = Field.class.getDeclaredField("modifiers");
+        modifiersField.setAccessible(true);
+        modifiersField.setInt(sqlParserExecutorFiled, sqlParserExecutorFiled.getModifiers() & ~Modifier.FINAL);
+        
+        Field modifiersField2 = Field.class.getDeclaredField("modifiers");
```
Comment (soulasuna) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:70]: Modify the extra blank lines.

**Review by soulasuna — COMMENTED** [review]
@zzq1314zll You have to look at the code specification first.
https://shardingsphere.apache.org/community/en/contribute/code-conduct/

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class SQLParserEngineTest {
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:37]: Hi @zzq1314zll 
Thanks for your first PR here. ;-) 
Could U define these test classes as final classes? As you know, these classes are bound to have no child classes extending them, aren't they?

Comment (tristaZero) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:37]: Generally LGTM.

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java (hunk 11 above)
Comment (zzq1314zll) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:37]: Okay thank you

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+                });
+        parseTreeCacheField.set(sqlParserEngine, parseTreeCache);
+        sqlParserEngine.parse(SQL, true);
+        verify(sqlParserExecutor, times(1)).parse(SQL);
+        sqlParserEngine.parse(SQL, true);
+        verify(sqlParserExecutor, times(1)).parse(SQL);
+        sqlParserEngine.parse(SQL, false);
+        verify(sqlParserExecutor, times(2)).parse(SQL);
+    }
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:73]: Hi @zzq1314zll Thanks for this Nice job. ;-)

Actually, we suggest our contributors remove redundant blank lines. If you really need a new line to split your programming, which means you need a new function to handle it. Therefore, this line 73 is expected to be removed.  Sometimes, developers will feel dull about these checkstyle or formatting issues, but please believe people all over the world will enjoy your elegant programming.

**Review by tristaZero — APPROVED** [review]

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java (hunk 12 above)
Comment (zzq1314zll) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/test/java/org/apache/shardingsphere/sql/parser/api/SQLParserEngineTest.java:73]: I may not be familiar with SS, the last test can not be completed, sorry！Wait until I get familiar with it before continuing.

**Pull request merged**